### DISCUSSION TITLE
Recommended way to get updated fields when listening to session events using do_orm_execute

### DISCUSSION THREAD

**Discussion by mtgraves** [sqlalchemy/sqlalchemy#11342] [Usage Questions] [answered]
When using `sqlalchemy` event listeners, I'm wondering the recommended/best way to get the actual updated values while using the `do_orm_execute` session event.

For example, when detecting ORM events using something like `before_flush` I can do something like:

```python
def _before_flush(session, flush_context, instances):
    for entry in session.dirty:
        # below would print the class instance of my updated table row
        print(entry)
event.listen(session, "before_flush", _before_flush)
```

I'm hoping to find a (at least nearly) analogous way of detecting updates that were made through `execute()` statements using `do_orm_execute`.

```python
def _receive_orm_execute(orm_execute_state: ORMExecuteState):
    if orm_execute_state.is_update:
        # what is the best way to grab updates here?
        # the following seems to have tuples with updated field values, but should I be using this:
        # orm_execute_state.execution_options['_sa_orm_update_options']._resolved_keys_as_propnames

event.listen(session, "do_orm_execute", _receive_orm_execute)
```

FWIW:  using `sqlalchemy` v1.4.x (due to project constraints) but will take insights from any version.

Thanks in advance.

**Comment by zzzeek** [comment]
ORMExecuteState has fields for everything that's going to happen including the statement that's going to run and the parameters. see https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState.statement  https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState.parameters etc.

**Reply by mtgraves** [reply]
@zzzeek Thank you for the reply!  I was originally hoping to find useful information in the `parameters` but that is coming back as `None` for me when i run an `update()`.  What should be there?

**Reply by zzzeek** [reply]
What kind of an update()?  if you have an `update().values()` and you aren't passing parameters separately to `session.execute()`, then there's no parameters.  you'd have to look inside the update construct itself inside `.statement`.     inspection of SQL statement objects is only semi-public, so you'd need to poke around the [Update](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/dml.py#L1509) construct to find what you need.

Bigger picture I usually would not recommend trying to pull out the parameters from an update object, since in particular for an ORM enabled statement there is not necessarily a 1-1 relationship between the parameters of an UPDATE and the model object, and would instead approach the problem differently (which I can't comment on because I dont know the actual thing you want to do).

**Reply by mtgraves** [reply]
your guess on use is correct - an example statement would be:

```python
session.execute(
    update(ExampleTable)
    .where(ExampleTable.id == 123)
    .values(
        field_1 = "stuff",
        field_2 = "things"
    )
)
session.commit()
```

In general, what I'm attempting to do is to detect the field_1 and field_2 updated values ("stuff" and "things") using an event listener.  I know I could just write the update statement not skipping the ORM and detect like in my first example in my original post, but this project is large and some of our devs use this pattern for updates.  Ignoring other project management/pipeline solutions for catching these, I'd like to allow them to continue to write these and still listen in on them and robustly detect those field changes.

Any suggestions are much appreciated!   I assumed there might be some gotchas hidden in here, hence the reason i posted for guidance.

**Reply by zzzeek** [reply]
right but why do you need to detect them?   you can use things like triggers etc.

**Reply by mtgraves** [reply]
Do you mean at the database level?  Like a [mariadb trigger](https://mariadb.com/kb/en/trigger-overview/)?  Or are you referring to something in SQLAlchemy that i've glossed over?

**Reply by zzzeek** [reply]
well like a real trigger in mariadb, sure. depends on what you actually are doing

**Reply by zzzeek** [reply]
if they are foreign key cols, you might use ON UPDATE, etc

**Reply by mtgraves** [reply]
part of the requirement is I need to be able to grab that information, construct dict representation of the updated row (so json serializable), then send off via http to another service along with some other metadata that comes from the service (fastapi).  So it makes sense to me to have this live in this layer of the stack

**Reply by zzzeek** [reply, marked as answer]
when you have your update() object, you can compile it to get the params portably:

```
>>> from sqlalchemy import *
>>> stmt = update(table('t', column('q'), column('z'))).values(q=5, z=7)
>>> stmt.compile().params
{'q': 5, 'z': 7}
```

another way to intercept at a low level but on the client side, if you can make that call very cheap and non-blocking you can stick it into an onupdate function like at https://docs.sqlalchemy.org/en/20/core/defaults.html#context-sensitive-default-functions  though that would be an unusual approach

using server side triggers you could have a trigger populate a separate table where you have a process that would kick off a job by scanning that table for new tasks.

**Reply by mtgraves** [reply]
interesting!  The more I learn about sqlalchemy the more i feel like i know nothing about sqlalchemy haha.

Couple of follow-up questions to your last answer:
(1) how would I get the model (`ExampleTable` from my ex above) from the result of `compile()`?
(2) could there ever be an update construct that has changes to more than one model/table?  I'm wondering because, for instance, the `orm_execute_state.execution_options['_sa_orm_update_options']._matched_objects` shows up as a list of model classes as opposed to a single one.  If so, how would that be represented in the `.compile().params` output?

**Reply by mtgraves** [reply]
ah for (1) above - is there any better place than in `.compile().statement.entity_description`?

**Reply by mtgraves** [reply]
@zzzeek one last question for you - where would be the best place to find the existing values for all of the other fields within the `OrmExecuteState` object?    I found them all the way down in `orm_execute_state.execution_options['_sa_orm_update_options']._matched_objects` but this feels like the totally wrong place to grab them from.

**Reply by zzzeek** [reply]
for entity description, that's on the statement itself, just use it directly  "context.statement.entity_description" 


for "other fields" I dont know what fields you are looking for.   `_sa_orm_update_options` is internal stuff

**Reply by mtgraves** [reply]
Apologies for the vague ask - by "other fields" i mean existing field values on a table that were not updated.  In your example above, imagine a third column "r" that had a value that was not updated as a part of the update() call.

More generally - how would I determine the values of that entity before the update?

**Reply by zzzeek** [reply]
you've lost me; we are talking about an update() statement with values() in it.  there's no "entity" in play here since you aren't using the unit of work that deals with specific objects.

**Reply by zzzeek** [reply]
`entity_description` is just the class that's the target of the update but doesnt correspond to a specific set of values

**Reply by mtgraves** [reply]
are the values stored anywhere in the OrmExecuteState?  I imagine I could just query the session for those, but if they're already there then i'll avoid that step.

**Reply by zzzeek** [reply]
the use case we're talking about from the beginning is this:

```py

with Session(e) as sess:
    stmt = update(MyModel).values(x=5, y=10)
    sess.execute(stmt)

```

the values you can get from that are `x=5` and `y=10`.      There's no "values that weren't updated" in this case.

**Reply by zzzeek** [reply]
unless you want to use RETURNING to get other columns from the updated rows?  you can do that, but that's not an event hook thing, that's getting the result back and consuming it

**Reply by mtgraves** [reply]
Can you please elaborate on that last statement?  Re: getting the result back and consuming it.

**Reply by zzzeek** [reply]
take a look at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-update-delete-and-custom-where-criteria
